feat: support phone based booking api

Udit-takkar opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes #XXXX (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description)

https://app.campsite.com/cal/posts/muplmwz5gjpg

1) Create an event type and go to advance settings and make email field as not required and Phone number required and visible on booking page.

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

request payload:-

```
{
  "start": "2024-11-18T05:15:00.000Z",
  "eventTypeId": 32,
  "attendee": {
    "name": "John Doe",
    "attendeePhoneNumber":"+919876543210",
    "timeZone": "Asia/Kolkata",
    "language": "en"
  }
}
```

![Screenshot 2024-11-14 at 6 41 38 PM](https://github.com/user-attachments/assets/24bfedf1-1416-485f-b9ba-cf6a252aa0cc)


## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

Udit-takkar committed:
chore: save progress

Udit-takkar committed:
feat: support phone based booking using API v2

Udit-takkar commented on packages/platform/types/bookings/2024-08-13/inputs/create-booking.input.ts:
```diff
@@ -44,6 +76,17 @@ class Attendee {
+  @Validate((value: string) => !value || isValidPhoneNumber(value), {
+    message: "Invalid phone number format. Please use international format.",
+  })
+  attendeePhoneNumber?: string;
```
`attendeePhoneNumber` is actually the reserved identifier for phone based booking.

supalarry replied:
We can receive it as `phoneNumber` and then convert it to `attendeePhoneNumber` within function `transformInputCreateBooking` in file `apps/api/v2/src/ee/bookings/2024-08-13/services/input.service.ts` - just need to update what you already have done to:
```
attendeePhoneNumber: inputBooking.attendee.phoneNumber
```

supalarry commented on packages/platform/types/bookings/2024-08-13/inputs/create-booking.input.ts:
```diff
@@ -28,13 +55,18 @@ class Attendee {
-  @IsEmail()
-  email!: string;
+  @RequireEmailOrPhone()
+  email = ""; // Default to empty string
```
I would not default to empty email but instead make it optional:
```
email?: string;
```

Udit-takkar replied:
Actually handleNewBooking expects email as string because right now in DB the expected type is string. for phone based booking we convert phone number to a placeholder email :- +91 98765432100 --> 919876543210@sms.cal.com

supalarry replied:
In such case why not still mark it optional but in the input service set it to empty string if it is not passed? The input is just an input it should not have any idea about what happens down the line but should rather reflect what data user can provide.

Udit-takkar replied:
done

supalarry commented on packages/platform/types/bookings/2024-08-13/inputs/create-booking.input.ts:
```diff
@@ -1,24 +1,51 @@
 import type { BookingLanguageType } from "./language";
 import { BookingLanguage } from "./language";
 
+function RequireEmailOrPhone(validationOptions?: ValidationOptions) {
```
We should also add e2e tests for booking with phone in `apps/api/v2/src/ee/bookings/2024-08-13/controllers/user-bookings.controller.e2e-spec.ts` or can copy test setup from the file and create a new `phone-bookings.e2e-spec.ts` file.

supalarry reviewed (dismissed):
The PR looks good but the problem I see is that right now we can make a booking using attendee phone number for event types that don't have attendee phone number booking field. Just like we have in this PR `RequireEmailOrPhone` we should have the same logic in event type.

The 4.5 release blog states
> To enable this (phone based bookings), go to your event type and enable 'Phone' under booking questions, mark it as required, and you can then turn off 'Email'.

So i think we should follow same logic and at least allow enabling phone based event types (event type without email but with phone) and allow an event type to have either phone or email booking field in files I described in the [thread](https://app.campsite.com/cal/posts/muplmwz5gjpg#comment-tbbjcorg5wjx).

However, to not block users who do have event types with phone I guess we could merge this after tests are added to unblock those users, but a follow up for the event types would be needed. Do you think you would have time for that?

Udit-takkar committed:
chore: change name of the variable

Udit-takkar committed:
fix: update event types and booking fields for improved validation and new properties

- Updated imports in input and output event types services to use the correct library version.
- Added new `avatarUrl` and `successRedirectUrl` properties to the Swagger documentation for user and booking schemas.
- Enhanced booking fields validation by changing the `required` property to a boolean type.
- Improved descriptions and examples in the OpenAPI specifications for better clarity.
- Refactored booking fields input and output classes to include new properties and ensure proper validation.

Udit-takkar committed:
fix: booking with phone number

Udit-takkar committed:
test: add event type update test

Udit-takkar committed:
test: add test for creating booking

Udit-takkar committed:
chore: update unit test

Udit-takkar committed:
fix: tests

Udit-takkar commented on apps/api/v2/src/ee/event-types/event-types_2024_06_14/services/input-event-types.service.ts:
```diff
@@ -22,7 +22,7 @@ import {
   SystemField,
   CustomField,
-} from "@calcom/platform-libraries";
+} from "@calcom/platform-libraries-0.0.0";
```
do i need to change version here?

Udit-takkar committed:
fix: add turbo

supalarry commented on apps/api/v2/src/ee/bookings/2024-08-13/controllers/e2e/team-bookings.e2e-spec.ts:
```diff
@@ -205,6 +206,83 @@ describe("Bookings Endpoints 2024-08-13", () => {
+            },
+          },
+          {
+            name: "email",
```
It says "phoneOnlyEventTypeId" but here we also have email booking field. Shouldn't we remove email booking field because we have "attendeePhoneNumber".

supalarry commented on apps/api/v2/src/ee/bookings/2024-08-13/controllers/e2e/team-bookings.e2e-spec.ts:
```diff
@@ -322,6 +400,61 @@ describe("Bookings Endpoints 2024-08-13", () => {
+          attendee: {
+            name: "alice",
+            phoneNumber: "+919876543210",
+            email: "test@test.com",
```
Also here - if we are testing a phone based booking shouldn't we remove `email` from the payload?

supalarry commented on apps/api/v2/src/modules/organizations/controllers/event-types/organizations-event-types.e2e-spec.ts:
```diff
@@ -577,6 +577,47 @@ describe("Organizations Event Types Endpoints", () => {
+      const body: UpdateTeamEventTypeInput_2024_06_14 = {
+        bookingFields: [
+          {
+            type: "email",
```
Like before - shouldn't we also omit email booking field here?

Udit-takkar committed:
Merge branch 'main' into feat/phone-based-api

supalarry commented on packages/platform/types/event-types/event-types_2024_06_14/inputs/booking-fields.input.ts:
```diff
@@ -58,6 +58,11 @@ export class EmailFieldInput_2024_06_14 {
+  @IsBoolean()
+  @IsOptional()
+  @DocsProperty()
+  required = true;
```
Why does it have to be required if we can have phone based bookings aka without email?

supalarry replied:
In the request payload of the PR description there is only attendee phone number.

Udit-takkar replied:
there was no required field defined for email i just set the default value true if it hasn't been passed as false or empty

Udit-takkar committed:
chore: remove emial

Udit-takkar committed:
refactor: make email optional and move it to service

Udit-takkar committed:
fix: test

supalarry committed:
chore: bump platform-libraries

Udit-takkar committed:
fix: e2e test
